Lucene Index Settings (Analyzers)

Skrypt opened the pull request:
TODO :  
- [x] make indexSettingsList in LuceneIndexingService thread safe.
- [x] Make analyzers name configurable on the service registration so that we can discover the registered ones without listing services registered by the DI. See Localization Middleware `RequestLocalizationOptions`.
- [x] ~~Save settings in database with a proper driver.~~ Won't do on this PR.
- [x] Make rebuild/reset work on Azure (request timeout) issue. **Mitigated with performance fix**

Some for later : 

- [x] ~~Tokenized and Sanitized~~ Won't do on this PR.

Fixes #3165, #2717, #4624, #4566, #3497

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -0,0 +1,209 @@
+
+        public ILogger Logger { get; }
+
+        public async Task ProcessContentItemsAsync(string indexName = default)
```
This was not processing one index at a time. I kept the option but now we can pass the indexName to start processing a specific index.

jtkech replied:
Hmm, we would need to check again how indexing sates are working (e.g last task ids in `lucene.status.json`), but if only one index has been rebuilt, not sure that it was doing extra work related to the other ones.

Skrypt replied:
That's what I experienced. I need to retry, maybe it was because of the Lucene Background worker. But you can see in the code that it loops on all indexes.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -0,0 +1,209 @@
+                    // Update task ids
+                    lastTaskId = batch.Last().Id;
+
+                    foreach (var index in allIndices)
```
Here it will set the LastTaskId of all indices once it completed it's task of indexing all content items.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -0,0 +1,209 @@
+                            // Update the document from the index if its lastIndexId is smaller than the current task id. 
+                            await indexHandlers.InvokeAsync(x => x.BuildIndexAsync(context), Logger);
+
+                            foreach (var index in allIndices)
```
Here it loops on all indices

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/LuceneIndexingService.cs:
```diff
@@ -50,21 +53,28 @@ public class LuceneIndexingService
 
-            // Find the lowest task id to process
-            var lastTaskId = int.MaxValue;
-            foreach (var indexName in _indexManager.List())
```
Here it was returning all indices by looking at folders located in the /App_Data/Lucene/ folder and return a list of folders name. So every time ProcessContentItemsAsync was called it was indexing all indices.

Skrypt commented:
Here are some tests I made with Analyzers.
I created a custom analyzer based on the FrenchAnalyzer that is provided by Lucene.NET and added a ASCIIFoldingFilter to test if it would order the words with diacritics properly.

Here is the result with the standard analyzer : 

![screencapture-localhost-44300-OrchardCore-Lucene-Admin-Query-2019-06-23-00_33_16](https://user-images.githubusercontent.com/3228637/59971672-93baee80-954e-11e9-9c41-837331bd0b46.png)

And here is the result with my custom french analyzer : 

![screencapture-localhost-44300-OrchardCore-Lucene-Admin-Query-2019-06-23-00_34_24](https://user-images.githubusercontent.com/3228637/59971674-adf4cc80-954e-11e9-8634-8d96e139163a.png)

Though, something not expected is that if I'm trying to order by ContentItem.DisplayText with the french analyzer it doesn't work.

![screencapture-localhost-44300-OrchardCore-Lucene-Admin-Query-2019-06-23-00_37_44](https://user-images.githubusercontent.com/3228637/59971702-44c18900-954f-11e9-9a69-b5c336a41560.png)

Skrypt commented:
So far I've had quite different results while using different fields. I need to test more to make sure it works properly. Sometimes doing a publish on a content item will fix it's sort order because it will force it to be analyzed with the current index Analyzer. Doing a rebuild doesn't work all the time for that matter (need to find why).

Skrypt commented:
In DefaultContentIndexHandler we are always indexing as only "Stored" while in AspectsContentIndexHandler we are indexing as "Stored" but also adding a second index field as "Analyzed". So I think we should maybe do the same in DefaultContentIndexHandler.

Skrypt commented:
Here is an other test. While using a TitlePart that is "Stored" and also "Analyzed" I get different results than when I'm setting 2 different fields for them.

Right now we set TitlePart as Stored and Analyzed but when we do a SORT query on a Stored field it will returns this : 

![screencapture-localhost-44300-OrchardCore-Lucene-Admin-Query-2019-06-23-04_30_47](https://user-images.githubusercontent.com/3228637/59973565-b1e51680-956f-11e9-893e-639c2a6b3b4e.png)

Though if I set a TitlePart.Analyzed field which is only analyzed I get this : 

![screencapture-localhost-44300-OrchardCore-Lucene-Admin-Query-2019-06-23-04_25_44](https://user-images.githubusercontent.com/3228637/59973580-e6f16900-956f-11e9-8c57-fee153b3146a.png)

Notice that the second one is properly ordering the results.

agriffard commented:
This is not working after a setup of a new Blog:
```
ArgumentNullException: Value cannot be null.
Parameter name: key
System.Collections.Generic.Dictionary<TKey, TValue>.FindEntry(TKey key)

OrchardCore.Lucene.Services.LuceneAnalyzerManager.CreateAnalyzer(string name) in LuceneAnalyzerManager.cs
+
            if (_analyzers.TryGetValue(name, out var analyzer))
```

This is not working on an existent instance of a blog:
```
NullReferenceException: Object reference not set to an instance of an object.
OrchardCore.Lucene.Controllers.AdminController.Edit(string indexName) in AdminController.cs, line 108
```

I can just create and edit a new index.

Skrypt commented:
Yes, I need to fix Unit Tests to create a default index. I removed that code as I didn't liked how it was adding/removing and adding back settings.

Skrypt commented:
@agriffard Ok it's fixed. It was an issue with the recipe. I updated the contentStep.cs to take proper index settings and updated the TheBlog recipe.

agriffard commented:
OK for the recipe.
How can I select another analyzer than the standard one?

Skrypt commented:
I added a FrenchAnalyzer project in the solution that you need to enable.

agriffard commented:
Thank you.
Some documentation about indexing would great too.

The previous Orchard meeting was interesting because many concepts were explained during it.

Skrypt commented:
Yes, I agree, we need to make things easier related with Indexing.

Skrypt commented:
I will now work on indexing drafts. Should be pretty trivial.

Skrypt commented:
![image](https://user-images.githubusercontent.com/3228637/63214397-b64c2c80-c0e5-11e9-8eff-69b6460ba60b.png)

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Lucene.FrenchAnalyzer/Startup.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public override void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<LuceneOptions>(o =>
```
Good. All the existing analyzers should be registered in the Lucene module.

Skrypt replied:
Yes, still need to add that.

Skrypt replied:
@jtkech We need to be able to do here the same as with the LocalizationMiddleware. We need to have a method to do AddAnalyzer((string)analyzerName). And probably a method to get All providers without requiring to get them from the DI each time.

jtkech replied:
Okay as you want, here the code of `AddSupportedCultures` in `RequestLocalizationOptions.cs`.

        public RequestLocalizationOptions AddSupportedCultures(params string[] cultures)
        {
            // But i think that you don't want to loose the previuos ones.
            var supportedCultures = new List<CultureInfo>();

            foreach (var culture in cultures)
            {
                supportedCultures.Add(new CultureInfo(culture));
            }

            SupportedCultures = supportedCultures;
            return this;
        }

sebastienros commented:
Not sure if I mentioned that already but we should also change the settings to be stored in the database, at least handle it in the distributed branch, such that when we create an index it can be replicated on all nodes. The state of the indexation has to remain on the FS though.

I assume you'll remove the FrenchAnalyzer project before we merge?

Skrypt commented:
Ok for settings. You sure though because Lucene indexes are stored locally already. I never thought we would want to use it as a distributed service.

Yes for the FrenchAnalyzer.

sebastienros commented:
We'll change the settings later, probably in the distributed branch.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/LuceneIndexingService.cs:
```diff
@@ -87,14 +113,31 @@ public async Task ProcessContentItemsAsync()
+                    var contentManager = scope.ServiceProvider.GetRequiredService<IContentManager>();
+                    var indexHandlers = scope.ServiceProvider.GetServices<IContentItemIndexHandler>();
+
+                    var contentItems = await contentManager.GetAsync(batch.Select(x => x.ContentItemId));
```
Here is what I'm considering right now.

```
var contentItems = await contentManager.GetAsync(batch.Select(x => x.ContentItemId), indexSettings.IndexDrafted);
```

So this would return all the "latest" content items. We should maybe consider changing the (bool)IndexDrafted to (bool)IndexLatest.

sebastienros replied:
Looks good

sebastienros commented:
Thinking about why I was iterating over all indices, the logic is that while we process a content item and get its fields, we can update every index at the same time instead of reading the content item multiple times for each index.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -87,40 +114,50 @@ public async Task ProcessContentItemsAsync()
 
-                        foreach (var index in allIndices)
+                    foreach (var indexSettings in indexSettingsList)
+                    {
```
We are still iterating on each indices I just moved it higher to be able to get the IndexSettings.LatestIndex for the contentManager query made at line 127.

Skrypt replied:
I understand that the logic is that the Lucene Background worker doesn't matter about which index to process it does them all.

sebastienros replied:
Good then, thanks

sebastienros commented on src/OrchardCore/OrchardCore.Mvc.Core/Utilities/StringExtensions.cs:
```diff
@@ -236,6 +236,31 @@ public static string RemoveDiacritics(this string name)
+                }
+            }
+
+            return stringBuilder.ToString().ToLower();
```
Should this method do that? Looks like orthogonal to me.

Skrypt replied:
You are right I'm removing the ToLower()

Skrypt commented:
ToDo : Make sure NumericField has the proper length when we are indexing if the NumericField is set to scale = 0 (Int32?) or else (decimal?).

Skrypt commented:
Note to myself.
Make sure that ContentPickerFields sets a NULL value when there's no item selected.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Contents/Indexing/AspectsContentIndexHandler.cs:
```diff
@@ -36,6 +38,16 @@ public async Task BuildIndexAsync(BuildIndexContext context)
                 context.ContentItem.DisplayText,
                 DocumentIndexOptions.Store);
 
+            context.DocumentIndex.Set(
```
Is it only for ordering?

Skrypt replied:
Yes, but it should be removed in favor of the Tokenized field implementation. The Liquid filters should then allow to do ReplaceDiacritics.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -50,22 +53,46 @@ public class LuceneIndexingService
-            foreach (var indexName in _indexManager.List())
+            if (String.IsNullOrEmpty(indexName))
             {
+                indexSettingsList = _luceneIndexSettingsService.List();
```
This can cause issue if someone is removing or adding an index while we are indexing.

jtkech replied:
You could use an immutable array in `LuceneIndexSettingsService` without any locking unless for writing the file.

Skrypt replied:
Do it if you have time 😉 any help is appreciated at this point I'm way over my planning time on this 😄

jtkech replied:
Okay as i have time, remind me if i forget it ;)

Skrypt commented:
I'm putting this dev on ice for this week I need to finish some other things.

Skrypt commented:
TODO :  
- [ ] Save settings in database with a proper driver.
- [ ] Clean unescessary code.
- [ ] Make analyzers name configurable on the service registration so that we can discover the registered ones without listing services registered by the DI. See Localization Middleware RequestLocalizationOptions.
- [ ] Use ImmutableDictionnary to make IndexSettingsList thread safe.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -0,0 +1,249 @@
+
+            if (String.IsNullOrEmpty(indexName))
+            {
+                indexSettingsList = _luceneIndexSettingsService.List().Where(x => x.IndexInBackgroundTask).ToImmutableDictionary(x => x.IndexName, x => x);
```
Here @jtkech. It's working like a charm.

jtkech replied:
Yes, i assume because `LuceneIndexSettingsService` is a tenant singleton and then its `List()` returns a shared collection that might be mutated while iterating it.

So, here you would just need an isolated collection e.g by using a `ToList()`

    _luceneIndexSettingsService.List().ToList.Where( ... ).ToDictionary(x => x.IndexName, x => x);

But to not have to do the same elsewhere and because i think `ToList()` and `Where` (that use an enumerator) are not thread safe themselves, i think it is better to define in `LuceneIndexSettingsService` its **`_indexSettings` as an immutable array**. So that here you could just do

    _luceneIndexSettingsService.List().Where( ... ).ToDictionary(x => x.IndexName, x => x);

Then in `LuceneIndexSettingsService` you could do e.g

        public void CreateIndex(LuceneIndexSettings settings)
        {
            _indexSettings = _indexSettings.Add(settings);
            Update();
        }

Skrypt replied:
I think you are right because we could want to add an index while indexing an other one. Though I've had no issue with this scenario as I tested it. Every time we call the service it will create a new ImmutableDictionary from the ToList() call.

Skrypt replied:
I will try your suggestion as it would be cleaner.

sebastienros commented:
Fixes #3278

Skrypt commented:
Should fix #3497

Skrypt commented:
Why is this not done yet? @jtkech you didn't fix my bugs over night? I thought I would wake up and that it be fixed magically. 🦄 🐰 🌟

jtkech commented:
@Skrypt you have to do it yourself ;)
Seriously, let me know if you have any problem

Skrypt commented:
omg all alone right 😢

Skrypt commented:
LuceneIndexManager Line 89. This could potentially fail :  

```C#
writer.DeleteDocuments(contentItemVersionIds.Select(x => new Term("Content.ContentItem.ContentItemVersionId", x)).ToArray());
```

Skrypt commented:
Indexing 4812 content items in 20 seconds.

Skrypt commented:
One issue I found. When we modify a content item if the ModelState has an error the LuceneIndexingContentHandler still pass in VersionedAsync and UpdatedAsync which cause indexing a content item that is not yet persisted in the database.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs:
```diff
@@ -106,57 +136,78 @@ public async Task ProcessContentItemsAsync()
 
                     // Delete all the existing documents
-                    foreach (var index in allIndices)
+                    foreach (var index in updatedDocumentsByIndex)
```
@sebastienros please see this method instead of doing a foreach on AllIndices which is not required here.

Skrypt commented:
@sebastienros I think we could easily merge this PR just so that we could have the benefit of the fixes introduced by this PR. The remaining work could be easily done later and it would not affect at all the feature itself. Let me know if you see something that is wrong with this PR.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs:
```diff
@@ -514,6 +514,9 @@ private async Task<IActionResult> EditPOST(string contentItemId, string returnUr
 
+            contentItem = await _contentManager.GetAsync(contentItemId, VersionOptions.DraftRequired);
+            model = await _contentItemDisplayManager.UpdateEditorAsync(contentItem, this, false);
+
```
@jtkech If you have time please review this. I think this could be done better than that but I didn't want to fix the DefaultContentManager in this PR.

jtkech replied:
Okay i will do asap, i need to read your related comment.

Skrypt replied:
Yes, so far I think it does it 3 times which is wrong. But previously it was doing it 2 times when calling UpdatedAsync and PublishedAsync. It will remove the published item with the draft if you select the option to do so on the index setting.

jtkech commented:
@Skrypt some thoughts

- I saw on gitter about the session `IsCancelled` to prevent some indexings

- Also, e.g on publishing, to prevent the indexing to be done multiple times through `VersionedAsync`, `UpdatedAsync` and `PublishedAsync`, because **`LuceneIndexingContentHandler`** is a scoped service, i think **you could hold a simple list of item ids to check if an indexing has already been done**.

    Note: Idem for **`CreateIndexingTaskContentHandler`**.

- This not thread safe to enumerate a collection that is hold by a tenant singleton, even if we use a lock when writing it. There is the same "problem" with autoroute entries and localization entries for its root dictionaries. **At least in `LuceneIndexSettingsService.List()` you could just use a `.ToArray()`**, this even if `Linq` is not thread safe itself when enumerating the source, but creating a new collection is faster than enumerating for doing stuff, so that would reduce the chance to have a thread issue.

- Finally i also thought about distributed services and how to invalidate things that are cached at the tenant singleton level. But here also same "problem" with autoroute and localization entries, so maybe part of another PR e.g when i will re-work on distributed services.

    Note: Anyway Lucene is not tailored for distributed env ;) I was thinking about `OC.Indexing` but just checked that this module doesn't hold any data that would need to be in sync.

Skrypt commented:
This is what I was worried about. Session is scoped and thus it would be harder to determine if the current indexing relates to a controller that simply has a ModelState failing. I'm thinking about this and different solutions. One that comes to my mind is that YesSQL could raise events itself that the LuceneIndexingHandler could use instead of the ContentBaseHandler ones. Also, the EditPost method of the AdminController of OrchardCore.Contents expose a fact that we are doing DB transactions still when doing UpdateEditorAsync() before knowing if the ModelState is valid.

Here is some analysis of what we do : 

```
        private async Task<IActionResult> EditPOST(string contentItemId, string returnUrl, bool stayOnSamePage, Func<ContentItem, Task> conditionallyPublish)
        {
            //Here we retreive the contentItem from DB and trigger no events.
            var contentItem = await _contentManager.GetAsync(contentItemId, VersionOptions.Latest);

            if (contentItem == null)
            {
                return NotFound();
            }

            if (!await _authorizationService.AuthorizeAsync(User, OrchardCore.Contents.Permissions.EditContent, contentItem))
            {
                return Unauthorized();
            }
            
            //Here we update the retreived from DB contentItem with the current changes we require but this will trigger a UpdatingAsync and UpdateAsync events. This means that the ContentItem here will be reindexed by Lucene before knowing if the ModelState of the current request is valid.
            var model = await _contentItemDisplayManager.UpdateEditorAsync(contentItem, this, false);
            if (!ModelState.IsValid)
            {
                //Here is where the data gets desync because we updated the latest content item Lucene document previously but we decide to cancel the _session.save() transaction.
                _session.Cancel();
                return View("Edit", model);
            }

            //Here I added a second event triggering to properly index the latest content item.
            //This time since we are passing VersionOptions.DraftRequired it will trigger VersionedAsync and VersioningAsync events. Which will index the latest and current content items.
            contentItem = await _contentManager.GetAsync(contentItemId, VersionOptions.DraftRequired);

            //Here we trigger UpdatingAsync and UpdatedAsync.
            model = await _contentItemDisplayManager.UpdateEditorAsync(contentItem, this, false);

            // The content item needs to be marked as saved in case the drivers or the handlers have
            // executed some query which would flush the saved entities inside the above UpdateEditorAsync.

            //Here we added this line to make sure we do a _session.save() but in the DefaultContentManager we also do it. But here some thoughts. If the _session.Save(contentItem) fails because of a network issue... we triggered already Lucene indexing and we could potentially have this content item missing in the persistent storage DB.
            _session.Save(contentItem);

            await conditionallyPublish(contentItem);

            if (returnUrl == null)
            {
                return RedirectToAction("Edit", new RouteValueDictionary { { "ContentItemId", contentItem.ContentItemId } });
            }
            else if (stayOnSamePage)
            {
                return RedirectToAction("Edit", new RouteValueDictionary { { "ContentItemId", contentItem.ContentItemId }, { "returnUrl", returnUrl } });
            }
            else
            {
                return LocalRedirect(returnUrl);
            }
        }
```

jtkech commented:
as a reminder see my las comment in #4669 related to an sqlite locking issue on publishing + indexing and where i suggest: `one solution would be to do this indexing work in a deferred task if possible`

jtkech commented:
After taking another look on it, some thinkings.

- For `LuceneIndexingService.List()` you could use `.ToArray()` on the singleton service side, that would be more thread safe. And why not just declaring

      public LuceneIndexSettings[] List() => _indexSettings.ToArray();

    And then remove all the `ToArray()` in the consumer side if not necessary.

Will see other details, here the main points i thought about

- I think i understand the goal of `UpdatedAsync()` but i don't understand yet the goal of `VersionedAsync()`. Note: it has not been added in `CreateIndexingTaskContentHandler`.

- E.g `indexSettings.IndexLatest` is checked in `LuceneIndexingContentHandler` but not in `CreateIndexingTaskContentHandler` and / or in `ProcessContentItemsAsync()`.

- Do we really need all the stuff of `CreateIndexingTaskContentHandler`, `IndexingTaskManager`and `IndexingBackgroundTask`? Knowing that indexes are updated through content handlers and we can build / reset them through the admin.

- Finally i suggest that the content handler enlist deferred tasks by taking care to not enlist 2 same tasks for the same contentItemId (easy to do). This would also fix #4669 by updating the indexes (that may trigger an `SqlQuery`) in a subsequent scope.

jtkech commented:
@Skrypt

Some thinkings, not on skipe because of the big power outage in Quebec, i hope it will be OK soon ;)

- I worked on other things but i think i found a way to prevent multiple indexing through the content handler. This by using a deferred task, not because the other issue, but because it would be easy to enlist / remove lucene tasks through the handler to only keep the last one in a list that we pass to the deferred task. We already do this kind of things in `IndexingTaskManager` where i already fixed a little issue related to how we pass the list var to the deferred task. I will work on it soon.

- The last problem would be when the validation fails, but with the above pattern, with a simple "event" it would be simple to clear the list of lucene tasks, as a rollback.

- But sorry, i don't remember why we have the option to index the latest even if it is not published (a draft), you said to me because it doesn't have the same version id, but i still don't remember why it would be useful to index it. That said, if it is needed i think that just adding `UpdatedAsync()` is enough, not `VersionedAsync()`.

- Note: Yes, before the indexing was only done on `RemovedAsync` and on `PublishedAsync()` that is only triggered if the model is valid.

jtkech commented:
@Skrypt i merged dev and fixed the conflicts.

Then, if you want, i will reuse an `ImmutableList` for the lucene settings. We will still need a lock e.g when adding an entry because an immutable list doesn't guarantee atomicity. If 2 concurrent  threads  add an entry on the same source, before any entry has already been added, the last one that completes its operation update the source with a copy where only its entry has been added, the last wins and only one entry is added.

But an immutable collection is still useful e.g to provide a collection to a consumer, knowing that with a regular list, even we do a `ToList()` to provide a copy to the consumer, this is not thread safe because `Linq` is not thread safe itself.

Skrypt commented:
I'm wokring on completing this today.

Skrypt commented:
Ah yeah I probably had the same changes locally.

jtkech commented:
Oups sorry, didn't see that you started to re-work on it

I have some last changes to do tonight asap, then i will let you pick up what you want

i didn't re-try anything ;)

jtkech commented:
I did some last refactorings, not tested.

 **The following is just for infos.**

- **One last thing would be how we are caching / storing data**, we would need to use the `Load / Get pattern` and other caching / tokens rules. E.g `Get` data from a tenant level cache for read only, `Load` isolated data from the store for a mutation, when updating the store invalidate the cache so that on the next cache `Get` new isolated data are retrieved from the store to re-init the cache, and so on.

    This allows to rely on the data that are really persisted and, in a distributed env, an instance updating some data will be able, through a distributed signal, to invalidate the local caches of  other instances.

    We know that Lucene is not intended to be used with multiple instances sharing the same file system, otherwise maybe useful to have distributed events e.g to rebuild an index, if so we would have to do the above to keep settings in sync across all instances. Hmm, but if in one server the app was not running, its index files would have not be updated, so we will see.

- **Anyway, we will see in another PR** because we already have the same kind of concerns e.g with autoroute and localization entries.

    Note: We have many settings that don't follow the above pattern, but many of them need a tenant reload (but here not for the same reason, because some startups need to be re-run), so the settings will be reloaded from the related sources. So here in a distributed env we will just need a tenant reload distributed signal so that all instances will be in sync by reloading the settings from shared sources (e.g database).

Skrypt commented:
Related #4929

agriffard commented:
Related to #2717

PiemP commented:
Hi @Skrypt ,

There is an issue with the clause "minimum_should_match". It involves the "boost" clause. I suppose is necessary to fix the class BooleanQueryProvider.cs in the OrchardCore.Lucene.Core module from this: 
```
switch (property.Name.ToLowerInvariant())
                {
                    case "must":
                        occur = Occur.MUST;
                        break;
                    case "mustnot":
                    case "must_not":
                        occur = Occur.MUST_NOT;
                        break;
                    case "should":
                        occur = Occur.SHOULD;
                        break;
                    case "boost":
                        boolQuery.Boost = query.Value<float>();
                        break;
                    case "minimum_should_match":
                        boolQuery.MinimumNumberShouldMatch = query.Value<int>();
                        break;
                    default: throw new ArgumentException($"Invalid property '{property.Name}' in boolean query");
                }

switch (property.Value.Type)
                {
                    case JTokenType.Object:

                        boolQuery.Add(builder.CreateQueryFragment(context, (JObject)property.Value), occur);

                        break;
                    case JTokenType.Array:
                        foreach (var item in ((JArray)property.Value))
                        {
                            if (item.Type != JTokenType.Object)
                            {
                                throw new ArgumentException($"Invalid value in boolean query");
                            }
                            boolQuery.Add(builder.CreateQueryFragment(context, (JObject)item), occur);
                        }
                        break;
                    default: throw new ArgumentException($"Invalid value in boolean query");

                }

```

to this:
```
switch (property.Name.ToLowerInvariant())
                {
                    case "must":
                        occur = Occur.MUST;
                        break;
                    case "mustnot":
                    case "must_not":
                        occur = Occur.MUST_NOT;
                        break;
                    case "should":
                        occur = Occur.SHOULD;
                        break;
                    case "boost":
                        boolQuery.Boost = property.Value.Value<float>();
                        break;
                    case "minimum_should_match":
                        boolQuery.MinimumNumberShouldMatch = property.Value.Value<int>();
                        break;
                    default: throw new ArgumentException($"Invalid property '{property.Name}' in boolean query");
                }

switch (property.Value.Type)
                {
                    case JTokenType.Object:

                        boolQuery.Add(builder.CreateQueryFragment(context, (JObject)property.Value), occur);

                        break;
                    case JTokenType.Array:
                        foreach (var item in ((JArray)property.Value))
                        {
                            if (item.Type != JTokenType.Object)
                            {
                                throw new ArgumentException($"Invalid value in boolean query");
                            }
                            boolQuery.Add(builder.CreateQueryFragment(context, (JObject)item), occur);
                        }
                        break;
                    case JTokenType.Integer: 
                    case JTokenType.Float:
                        break;//boost or minimum_should_match values
                    default: throw new ArgumentException($"Invalid value in boolean query");

                }
```

Sorry but I don't know how to push a commit on your pull request. 😅

Skrypt commented:
Related https://github.com/OrchardCMS/OrchardCore/issues/4952

Skrypt commented:
@PiemP To do a commit on top of this PR you can't. You need to create a PR that targets skrypt/lucene-analyzers branch. Then I will review the PR and merge it on top of that branch and it will be added on top of this PR afterward.

Skrypt commented:
Please provide the ElasticSearch Query that you used so that I can see what's the difference. I know that in some cases the boost param needs to be set differently than what is documented in the ElasticSearch documentation.

PiemP commented:
> Please provide the ElasticSearch Query that you used so that I can see what's the difference. 

I have tested it with this query: 
```
{
    "query": {
        "bool":{
            "must":{
                "term": {
                    "Content.ContentItem.ContentType": "News"
                }
            }
        },
        "bool": {
            "should": [
                {
                    "bool":{
                        "must":{
                            "range": {
                                    "News.datadiscadenza": {
                                        "lte": "{{ "now" | date: "%Y%m%d%H%M" }}"
                                    }
                            }
                        }
                    }
                },
                {
                    "bool":{
                        "must":{
                            "term": {
                                "News.datadiscadenza": "NULL"
                            }
                        }
                    }
                }
            ],
            "minimum_should_match" : 1,
            "boost" : 1.0
        }
    },
    "sort": {
        "News.datadipubblicazione": {
            "order": "desc"
        }
    },
    "size": 3
}
```

> I know that in some cases the boost param needs to be set differently than what is documented in the ElasticSearch documentation.

I check it and I try to do a PR over your PR.

PiemP commented:
> I know that in some cases the boost param needs to be set differently than what is documented in the ElasticSearch documentation.

I have make some search around the web and I don't find information about boost value in a different mode than a float. I' have looked at the source code and it's always defined as a float property.  

Probably I miss something.

Skrypt commented:
What I meant is that by looking at your PR I see that you are doing : `property.Value.Value<int>()` which means that you require to get the `boost` value in a sub JToken. I will check you PR today.

PiemP commented:
Hi @Skrypt, 

Sorry but my family have drain my energy...A crazy weekend when Christmas is coming.

The first query I have sent to you is right, I have create it after I have see this example https://www.elastic.co/guide/en/elasticsearch/reference/7.5/query-dsl-bool-query.html . The "minimum_sould_match" clause and the "boost" clause can be used inside a bool query.

> What I meant is that by looking at your PR I see that you are doing : property.Value.Value<int>()

Yes, I have added it to the "minimum_sould_match" clause and the "boost" clause because I have looked at the others *QueryProvider classes and you use a code like that. And in the next switch code I have added a case to ignore value types like Integer and Float, to avoid the thrown of the exception made by this code:
```
default: throw new ArgumentException($"Invalid value in boolean query");
```

Skrypt commented:
Ok tested the changes from @jtkech and everything is working fine. The PR is ready.

Skrypt commented:
@PiemP I added some of your suggested changes. 
I removed this part which was doing nothing : 

```
                    case JTokenType.Integer: 
                    case JTokenType.Float:
                        break;//boost or minimum_should_match values
```

PiemP commented:
> @PiemP I added some of your suggested changes.
> I removed this part which was doing nothing :
> 
> ```
>                     case JTokenType.Integer: 
>                     case JTokenType.Float:
>                         break;//boost or minimum_should_match values
> ```

@Skrypt You have test your modify with my query? 
Because without this line of code the BooleanQueryProvider class throw this issue:
```
default: throw new ArgumentException($"Invalid value in boolean query");
```
This happen because the first switch parse the value to see which kind of clause is, and the second switch (where you have removed the line of code) is used to call a class that manage the clause: if is it a complex clause like "must" or "must_not" or "should". But in the case of "minimum_should_match" and "boost" clause the value is an integer value or a float value that no need other works and this code allow the class to ignore this kind of values.

Skrypt commented:
Yes I tested it. Here is my test query : 

```
{
  "query": {
    "bool": {
      "should": [
        {
          "bool": {
            "must": {
              "range": {
                "Content.ContentItem.CreatedUtc": {
                  "lte": "{{ "now" | date: "%Y%m%d%H%M" }}"
                }
              }
            }
          }
        },
        {
          "bool": {
            "must": {
              "term": {
                "Content.ContentItem.CreatedUtc": "NULL"
              }
            }
          }
        }
      ]
    },
    "minimum_should_match": 1,
    "boost": 1
  }
}
```

Skrypt commented:
Oh but wait a second I think you might be right because now I don't see it getting applied. Strange.

Skrypt commented:
@PiemP Ok look at what I did. What you did kind of defeated the purpose of the second switch which is to validate the Boolean Query only. By adding the integer and float JToken.Type it was allowing to use such types for a Boolean Query. So I separated the boost and minimum_should_match from that validation.

Skrypt commented:
Need to fix import/export before merging.

Skrypt commented:
![image](https://user-images.githubusercontent.com/3228637/71315131-21a36380-2424-11ea-9f84-7c72812b7520.png)

Skrypt commented:
Ok I reworked everything that seemed to be a weak point in the last discussion we had on this PR. I had forgot to look at deployments and recipes properly. Now it's tested and works. Also added the culture selection setting on each index. Let me know if you guys see something else. Of course I won't fix the Azure instance issue since it's a different scope than what I intended to work on for now.

jtkech commented:
So, i updated `LuceneIndexSettingsService` to be distributed compatible.

**It may fail because i didn't try it ;)**

 - Use the `Get / Load` pattern, `Get` an isolated object for caching and as read only, `Load` an isolated object for updating. Doing this we don't need any more an immutable dictionary.

- Don't refresh the cache on updating, just save the document and invalidate the local cache with a signal token. The cache will be refreshed when the document is requested again, by a `Get`.

    Note: With a distributed signal token, the local cache of each running instances will be invalidated.

    Note: Here we don't use the memory cache, just a field of the service that is a singleton. And we don't use a **deferred** signal token because we use a new session that we commit just after saving.

- **I just need 2 confirmations**:

    In the singleton service, **do we still want to create a new session** in place of retrieving the current scoped one? Here i needed a 2nd refactoring e.g to load a doc to be updated and save it **without using the cache** and in the **same session** (so that we don't store a new doc).

    **Do we still want to not store the index names inside their related settings?** And then having to retrieve them from the dictionary keys?

Skrypt commented:
Ok now we need to change the recipes and deployment again I guess.

jtkech commented:
No, it is still a singleton service and still returns exactly the same things.

We just have to try it

Skrypt commented:
I'm just tired I thought I saw that the GetSettingsAsync was returning now the LuceneIndexSettings object directly but no. And it's still saved as a `Dictionary<string, LuceneIndexSettings>()` so I have nothing to change in the recipe or deployment code.

jtkech commented:
Normally nothing to change, same type in the database and so on.

We just need to re-try it a little, e.g if i made an obvious mistake ;)

sebastienros commented:
Last step to make it distributed, in `LuceneIndexSettingsDocument` we need to add a custom string containing a guid that will be changed any time an index is added or removed. This will let each instance know that it needs to check for new or deleted indices if the local guid doesn't match (in the same status document). Then for each index we can also store `LastReset` and `LastBuild` with guids too and the background task would then check for each index if an action has to be done. If both guids have changed then only a rebuild is necessary. Then the local state store the matching guids.

I will work on that. This should not change recipes at all.
